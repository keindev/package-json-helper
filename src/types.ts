import { Dependency } from './fields/Dependency';

export enum PackageManager {
  NPM = 'npm',
  Yarn = 'yarn',
}

export enum PackageType {
  Module = 'module',
  Commonjs = 'commonjs',
}

export enum StringProps {
  Browser = 'browser',
  Description = 'description',
  License = 'license',
  Main = 'main',
  Types = 'types',
}

export enum StringListProps {
  BundledDependencies = 'bundledDependencies',
  CPU = 'cpu',
  Files = 'files',
  Keywords = 'keywords',
  Man = 'man',
  OS = 'os',
  Workspaces = 'workspaces',
}

export enum StringMapProps {
  Scripts = 'scripts',
  Config = 'config',
  Engines = 'engines',
  Directories = 'directories',
}

export enum DependenciesMapProps {
  Dependencies = 'dependencies',
  DevDependencies = 'devDependencies',
  OptionalDependencies = 'optionalDependencies',
  PeerDependencies = 'peerDependencies',
}

export enum PackageDependency {
  // https://docs.npmjs.com/files/package.json#engines
  Engines = 'engines',
  // https://docs.npmjs.com/files/package.json#dependencies
  Dependencies = 'dependencies',
  // https://docs.npmjs.com/files/package.json#devdependencies
  DevDependencies = 'devDependencies',
  // https://docs.npmjs.com/files/package.json#peerdependencies
  PeerDependencies = 'peerDependencies',
  // https://docs.npmjs.com/files/package.json#optionaldependencies
  OptionalDependencies = 'optionalDependencies',
}

export enum PackageRestriction {
  // https://docs.npmjs.com/files/package.json#bundleddependencies
  BundledDependencies = 'bundledDependencies',
  // https://docs.npmjs.com/files/package.json#os
  OS = 'os',
  // https://docs.npmjs.com/files/package.json#cpu
  CPU = 'cpu',
}

export enum PackageDependencyChangeType {
  Added = 'added',
  Bumped = 'bumped',
  Changed = 'changed',
  Downgraded = 'downgraded',
  Removed = 'removed',
  Unchanged = 'unchanged',
}

export enum PackageChangeCompareResult {
  More = 1,
  Less = -1,
  Equal = 0,
}

export type Maybe<T> = T | undefined;
export type JSONValue = undefined | string | number | boolean | null | JSONObject | JSONValue[];
export type JSONObject = { [Key in string]?: JSONValue };
export type IStringProps = { [key in StringProps]?: string };
export type IStringListProps = { [key in StringListProps]?: Set<string> };
export type IStringMapProps = { [key in StringMapProps]?: Map<string, string> };
export type IDependenciesMapProps = { [key in DependenciesMapProps]?: Map<string, Dependency> };
export type IPackageProps = IStringProps & IStringListProps & IStringMapProps & IDependenciesMapProps;

export const PackageInstallCommandMap = {
  [PackageManager.NPM]: 'install',
  [PackageManager.Yarn]: 'add',
};

export const StringPropsMap = Object.values(StringProps);
export const StringListPropsMap = Object.values(StringListProps);
export const StringMapPropsMap = Object.values(StringMapProps);
export const DependenciesMapPropsMap = Object.values(DependenciesMapProps);

export interface IPackageChange {
  link?: string;
  name: string;
  type: PackageDependencyChangeType;
  value: {
    current?: string;
    previous?: string;
  };
}
