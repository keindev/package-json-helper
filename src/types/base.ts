export enum CompareResult {
  More = 1,
  Less = -1,
  Equal = 0,
}

export type Maybe<T> = T | undefined;
export type JSONValue = undefined | string | number | boolean | null | JSONObject | JSONValue[];
export type JSONObject = { [Key in string]?: JSONValue };
