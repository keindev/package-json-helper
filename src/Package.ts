import fs from 'fs';
import path from 'path';
import semver, { SemVer } from 'semver';

import { IBugsLocation, IFunding, IPerson, JSONObject } from './types';
import utils from './utils';

export class Package {
  /** Name of the package */
  name: string;
  /** Package version, parseable by [`node-semver`](https://github.com/npm/node-semver). */
  version: string;
  /** Package description, listed in `npm search`. */
  description?: string;
  /** Keywords associated with package, listed in `npm search`. */
  keywords: Set<string>;
  /** The URL to the package's homepage. */
  homepage?: string;
  /** The URL to the package's issue tracker and/or the email address to which issues should be reported. */
  bugs?: IBugsLocation;
  /** The license for the package. */
  license?: string;
  /** Author of the package */
  author?: IPerson;
  /** A list of people who contributed to the package. */
  contributors: IPerson[];
  /** A list of people who maintain the package. */
  maintainers: IPerson[];
  /** Describes and notifies consumers of a package's monetary support information. */
  funding: IFunding[];
  /** The files included in the package. */
  files: Set<string>;

  constructor(value: string | JSONObject) {
    const data =
      typeof value === 'string' || typeof value === 'undefined'
        ? (JSON.parse(fs.readFileSync(value ?? path.resolve(process.cwd(), 'package.json'), 'utf-8')) as JSONObject)
        : value;

    this.name = utils.getName(data.name);
    this.version = utils.getVersion(data.version);
    this.description = utils.getString(data.description, { fieldName: 'Description' });
    this.keywords = utils.getStringSet(data.keywords, { fieldName: 'Keywords' });
    this.homepage = utils.getHomePage(data.homepage);
    this.bugs = utils.getBugsLocation(data.bugs);
    this.license = utils.getString(data.license, { fieldName: 'License' });
    this.author = utils.getPerson(data.author, 'Author');
    this.contributors = utils.getPersons(data.contributors, 'Contributors');
    this.maintainers = utils.getPersons(data.maintainers, 'Maintainers');
    this.funding = utils.getFunding(data.funding);
    this.files = utils.getStringSet(data.files, { fieldName: 'Files' });
  }

  get prereleaseVersion(): string {
    return [...(semver.prerelease(this.version) ?? [])].join('.');
  }

  get semanticVersion(): SemVer {
    return semver.coerce(this.version) as SemVer;
  }

  get cleanVersion(): string {
    return this.semanticVersion.version;
  }

  get scope(): string | undefined {
    return this.name.match(/^@(?<scope>[\w.-]+)\/(?<name>[\w.-]+)/i)?.groups?.scope;
  }
}

export default Package;
