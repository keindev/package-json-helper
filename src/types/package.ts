export enum ManagerType {
  NPM = 'npm',
  Yarn = 'yarn',
}

export enum PackageType {
  Module = 'module',
  Commonjs = 'commonjs',
}

export enum Dependencies {
  // https://docs.npmjs.com/files/package.json#engines
  Engines = 'engines',
  // https://docs.npmjs.com/files/package.json#dependencies
  Dependencies = 'dependencies',
  // https://docs.npmjs.com/files/package.json#devdependencies
  DevDependencies = 'devDependencies',
  // https://docs.npmjs.com/files/package.json#peerdependencies
  PeerDependencies = 'peerDependencies',
  // https://docs.npmjs.com/files/package.json#optionaldependencies
  OptionalDependencies = 'optionalDependencies',
}

export enum Restriction {
  // https://docs.npmjs.com/files/package.json#bundleddependencies
  BundledDependencies = 'bundledDependencies',
  // https://docs.npmjs.com/files/package.json#os
  OS = 'os',
  // https://docs.npmjs.com/files/package.json#cpu
  CPU = 'cpu',
}

export enum ChangeType {
  Added = 'added',
  Bumped = 'bumped',
  Changed = 'changed',
  Downgraded = 'downgraded',
  Removed = 'removed',
  Unchanged = 'unchanged',
}

export interface IChange {
  link?: string;
  name: string;
  type: ChangeType;
  value: {
    current?: string;
    previous?: string;
  };
}
