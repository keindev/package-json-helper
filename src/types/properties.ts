import { Dependency } from '../fields/Dependency.js';

export enum StringProps {
  Browser = 'browser',
  Description = 'description',
  License = 'license',
  Main = 'main',
  Types = 'types',
}

export enum StringListProps {
  BundledDependencies = 'bundledDependencies',
  CPU = 'cpu',
  Files = 'files',
  Keywords = 'keywords',
  Man = 'man',
  OS = 'os',
  Workspaces = 'workspaces',
}

export enum StringMapProps {
  Scripts = 'scripts',
  Config = 'config',
  Engines = 'engines',
  Directories = 'directories',
}

export enum DependenciesMapProps {
  Dependencies = 'dependencies',
  DevDependencies = 'devDependencies',
  OptionalDependencies = 'optionalDependencies',
  PeerDependencies = 'peerDependencies',
}

export type IStringProps = { [key in StringProps]?: string };
export type IStringListProps = { [key in StringListProps]?: Set<string> };
export type IStringMapProps = { [key in StringMapProps]?: Map<string, string> };
export type IDependenciesMapProps = { [key in DependenciesMapProps]?: Map<string, Dependency> };
export type IPackageProps = IStringProps & IStringListProps & IStringMapProps & IDependenciesMapProps;
