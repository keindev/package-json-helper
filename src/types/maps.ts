import { DependenciesMapProps, StringListProps, StringMapProps, StringProps } from './properties.js';

export const StringPropsMap = Object.values(StringProps);
export const StringListPropsMap = Object.values(StringListProps);
export const StringMapPropsMap = Object.values(StringMapProps);
export const DependenciesMapPropsMap = Object.values(DependenciesMapProps);
