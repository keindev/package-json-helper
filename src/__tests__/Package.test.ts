import Package from '../Package';
import { DependenciesMapProps } from '../types';

const base = {
  name: 'test',
  version: '1.0.1',
  description: 'package description',
  keywords: ['test', 'jest'],
  homepage: 'https://github.com/owner/project#readme',
  bugs: {
    url: 'https://github.com/owner/project/issues',
    email: 'project@hostname.com',
  },
  license: '(MIT OR Apache-2.0)',
  author: {
    name: 'Barney Rubble',
    email: 'b@rubble.com',
    url: 'http://barnyrubble.tumblr.com/',
  },
  contributors: [
    {
      name: 'Barney Rubble',
      email: 'b@rubble.com',
      url: 'http://barnyrubble.tumblr.com/',
    },
  ],
  maintainers: [
    {
      name: 'Barney Rubble',
      email: 'b@rubble.com',
      url: 'http://barnyrubble.tumblr.com/',
    },
  ],
  funding: {
    type: 'individual',
    url: 'http://example.com/donate',
  },
  files: ['__tests__/*.*'],
  type: 'module',
  main: 'lib/index.js',
  types: 'lib/index.d.ts',
  bin: 'lib/index.js',
  man: './man/doc.1',
  directories: {
    lib: 'src/lib',
    bin: 'local/binaries',
    jars: 'java',
  },
  repository: {
    type: 'git',
    url: 'https://github.com/facebook/react.git',
    directory: 'packages/react-dom',
  },
  scripts: {
    prepare: 'command -o lib/ -c src/index.js',
    compress: '{{ run command to compress files }}',
  },
  config: {
    port: '8080',
  },
  dependencies: {
    foo: '1.0.0 - 2.9999.9999',
    bar: '>=1.0.2 <2.1.2',
    baz: '>1.0.2 <=2.3.4',
    boo: '2.0.1',
    qux: '<1.0.0 || >=2.3.1 <2.4.5 || >=2.5.2 <3.0.0',
    asd: 'http://asdf.com/asdf.tar.gz',
    til: '~1.2',
    elf: '~1.2.3',
    two: '2.x',
    thr: '3.3.x',
    lat: 'latest',
    dyl: 'file:../dyl',
  },
  devDependencies: {
    'super-script': '~1.6.3',
  },
  peerDependencies: {
    tea: '2.x',
    'soy-milk': '1.2',
  },
  peerDependenciesMeta: {
    'soy-milk': {
      optional: true,
    },
  },
  bundledDependencies: ['render-blender', 'super-streams'],
  optionalDependencies: {
    'some-pkg': '1.x',
  },
  engines: {
    node: '>=0.10.3 <15',
  },
  os: ['darwin', 'linux', '!win32'],
  cpu: ['x64', 'ia32', '!arm'],
  publishConfig: {
    'allow-same-version': true,
  },
  workspaces: ['./packages/*'],
};

const alternative = {
  ...base,
  name: '@scope/package',
  private: true,
  bugs: {
    url: 'https://github.com/owner/project/issues',
  },
  license: 'SEE LICENSE IN LICENSE_FILE',
  author: 'Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)',
  contributors: ['Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)'],
  maintainers: ['Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)'],
  funding: [
    {
      type: 'individual',
      url: 'http://example.com/donate',
    },
    'http://example.com/donateAlso',
    {
      type: 'patreon',
      url: 'https://www.patreon.com/my-account',
    },
  ],
  browser: 'lib/index.js',
  bin: {
    name: './path/to/program/index.js',
    shortname: './path/to/program/index.js',
  },
  man: ['./man/foo.1', './man/bar.1'],
  repository: 'gitlab:user/repo',
};

describe('Package', () => {
  const pkgBase = new Package(base);
  const pkgAlternative = new Package(alternative);

  it('Parse base', () => {
    expect(pkgBase.toString()).toMatchSnapshot();
  });

  it('Parse alternative', () => {
    expect(pkgAlternative.scope).toBe('scope');
    expect(pkgAlternative.toString()).toMatchSnapshot();
  });

  it('Check missing dependencies', () => {
    expect(
      pkgBase.getMissingDependencies(DependenciesMapProps.Dependencies, ['foo', 'bar', 'zip', 'zap'])
    ).toMatchObject(['zip', 'zap']);
  });

  it('Check wrong version dependencies', () => {
    expect(
      pkgBase.getWrongVersionDependencies(
        DependenciesMapProps.Dependencies,
        new Map([
          ['foo', '>2.9999.9999'],
          ['bar', '4.x'],
          ['boo', '>1.0.2'],
          ['none', '1.x'],
          ['thr', '^3.3.8 <3.4.x'],
        ])
      )
    ).toMatchObject(['foo', 'bar', 'boo']);
  });
});
