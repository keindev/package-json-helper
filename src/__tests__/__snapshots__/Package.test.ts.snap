// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Package Parse alternative 1`] = `
"{
  \\"name\\": \\"@scope/package\\",
  \\"version\\": \\"1.0.1\\",
  \\"description\\": \\"package description\\",
  \\"license\\": \\"SEE LICENSE IN LICENSE_FILE\\",
  \\"private\\": true,
  \\"homepage\\": \\"https://github.com/owner/project#readme\\",
  \\"author\\": \\"Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)\\",
  \\"contributors\\": [
    \\"Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)\\"
  ],
  \\"maintainers\\": [
    \\"Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)\\"
  ],
  \\"keywords\\": [
    \\"jest\\",
    \\"test\\"
  ],
  \\"funding\\": [
    {
      \\"url\\": \\"http://example.com/donate\\",
      \\"type\\": \\"individual\\"
    },
    \\"http://example.com/donateAlso\\",
    {
      \\"url\\": \\"https://www.patreon.com/my-account\\",
      \\"type\\": \\"patreon\\"
    }
  ],
  \\"types\\": \\"lib/index.d.ts\\",
  \\"browser\\": \\"lib/index.js\\",
  \\"type\\": \\"module\\",
  \\"files\\": [
    \\"__tests__/*.*\\"
  ],
  \\"workspaces\\": [
    \\"./packages/*\\"
  ],
  \\"directories\\": {
    \\"bin\\": \\"local/binaries\\",
    \\"jars\\": \\"java\\",
    \\"lib\\": \\"src/lib\\"
  },
  \\"imports\\": {
    \\"#dep\\": {
      \\"node\\": \\"dep-node-native\\",
      \\"default\\": \\"./dep-polyfill.js\\"
    },
    \\"#internal/*\\": \\"./src/internal/*.js\\"
  },
  \\"exports\\": {
    \\"node\\": {
      \\"import\\": \\"./feature-node.mjs\\",
      \\"require\\": \\"./feature-node.cjs\\"
    },
    \\"./feature\\": {
      \\"node\\": \\"./feature-node.js\\",
      \\"default\\": \\"./feature.js\\"
    },
    \\"default\\": \\"./feature.mjs\\"
  },
  \\"bin\\": {
    \\"name\\": \\"./path/to/program/index.js\\",
    \\"shortname\\": \\"./path/to/program/index.js\\"
  },
  \\"man\\": [
    \\"./man/bar.1\\",
    \\"./man/foo.1\\"
  ],
  \\"repository\\": \\"gitlab:user/repo\\",
  \\"config\\": {
    \\"port\\": \\"8080\\"
  },
  \\"publishConfig\\": {
    \\"allow-same-version\\": true
  },
  \\"os\\": [
    \\"!win32\\",
    \\"darwin\\",
    \\"linux\\"
  ],
  \\"cpu\\": [
    \\"!arm\\",
    \\"ia32\\",
    \\"x64\\"
  ],
  \\"engines\\": {
    \\"node\\": \\">=0.10.3 <15\\"
  },
  \\"scripts\\": {
    \\"compress\\": \\"{{ run command to compress files }}\\",
    \\"prepare\\": \\"command -o lib/ -c src/index.js\\"
  },
  \\"dependencies\\": {
    \\"asd\\": \\"http://asdf.com/asdf.tar.gz\\",
    \\"bar\\": \\">=1.0.2 <2.1.2\\",
    \\"baz\\": \\">1.0.2 <=2.3.4\\",
    \\"boo\\": \\"2.0.1\\",
    \\"dyl\\": \\"file:../dyl\\",
    \\"elf\\": \\"~1.2.3\\",
    \\"foo\\": \\"1.0.0 - 2.9999.9999\\",
    \\"lat\\": \\"latest\\",
    \\"qux\\": \\"<1.0.0 || >=2.3.1 <2.4.5 || >=2.5.2 <3.0.0\\",
    \\"thr\\": \\"3.3.x\\",
    \\"til\\": \\"~1.2\\",
    \\"two\\": \\"2.x\\"
  },
  \\"devDependencies\\": {
    \\"super-script\\": \\"~1.6.3\\"
  },
  \\"peerDependencies\\": {
    \\"soy-milk\\": \\"1.2\\",
    \\"tea\\": \\"2.x\\"
  },
  \\"peerDependenciesMeta\\": {
    \\"soy-milk\\": {
      \\"optional\\": true
    }
  },
  \\"bundledDependencies\\": [
    \\"render-blender\\",
    \\"super-streams\\"
  ],
  \\"optionalDependencies\\": {
    \\"some-pkg\\": \\"1.x\\"
  }
}"
`;

exports[`Package Parse base 1`] = `
"{
  \\"name\\": \\"test\\",
  \\"version\\": \\"1.0.1\\",
  \\"description\\": \\"package description\\",
  \\"license\\": \\"(MIT OR Apache-2.0)\\",
  \\"homepage\\": \\"https://github.com/owner/project#readme\\",
  \\"author\\": \\"Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)\\",
  \\"contributors\\": [
    \\"Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)\\"
  ],
  \\"maintainers\\": [
    \\"Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)\\"
  ],
  \\"keywords\\": [
    \\"jest\\",
    \\"test\\"
  ],
  \\"types\\": \\"lib/index.d.ts\\",
  \\"type\\": \\"module\\",
  \\"files\\": [
    \\"__tests__/*.*\\"
  ],
  \\"workspaces\\": [
    \\"./packages/*\\"
  ],
  \\"directories\\": {
    \\"bin\\": \\"local/binaries\\",
    \\"jars\\": \\"java\\",
    \\"lib\\": \\"src/lib\\"
  },
  \\"exports\\": \\"./lib/index.js\\",
  \\"bin\\": {
    \\"test\\": \\"lib/index.js\\"
  },
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"https://github.com/facebook/react.git\\",
    \\"directory\\": \\"packages/react-dom\\"
  },
  \\"config\\": {
    \\"port\\": \\"8080\\"
  },
  \\"publishConfig\\": {
    \\"allow-same-version\\": true
  },
  \\"os\\": [
    \\"!win32\\",
    \\"darwin\\",
    \\"linux\\"
  ],
  \\"cpu\\": [
    \\"!arm\\",
    \\"ia32\\",
    \\"x64\\"
  ],
  \\"engines\\": {
    \\"node\\": \\">=0.10.3 <15\\"
  },
  \\"scripts\\": {
    \\"compress\\": \\"{{ run command to compress files }}\\",
    \\"prepare\\": \\"command -o lib/ -c src/index.js\\"
  },
  \\"dependencies\\": {
    \\"asd\\": \\"http://asdf.com/asdf.tar.gz\\",
    \\"bar\\": \\">=1.0.2 <2.1.2\\",
    \\"baz\\": \\">1.0.2 <=2.3.4\\",
    \\"boo\\": \\"2.0.1\\",
    \\"dyl\\": \\"file:../dyl\\",
    \\"elf\\": \\"~1.2.3\\",
    \\"foo\\": \\"1.0.0 - 2.9999.9999\\",
    \\"lat\\": \\"latest\\",
    \\"qux\\": \\"<1.0.0 || >=2.3.1 <2.4.5 || >=2.5.2 <3.0.0\\",
    \\"thr\\": \\"3.3.x\\",
    \\"til\\": \\"~1.2\\",
    \\"two\\": \\"2.x\\"
  },
  \\"devDependencies\\": {
    \\"super-script\\": \\"~1.6.3\\"
  },
  \\"peerDependencies\\": {
    \\"soy-milk\\": \\"1.2\\",
    \\"tea\\": \\"2.x\\"
  },
  \\"peerDependenciesMeta\\": {
    \\"soy-milk\\": {
      \\"optional\\": true
    }
  },
  \\"bundledDependencies\\": [
    \\"render-blender\\",
    \\"super-streams\\"
  ],
  \\"optionalDependencies\\": {
    \\"some-pkg\\": \\"1.x\\"
  }
}"
`;
